lib/foreman_maintain/upgrade_runner.rb

```diff
@@ -57,25 +36,35 @@ def clear_current_target_version
 
     def initialize(version, reporter, options = {})
       super(reporter, [], options)
-      @tag = self.class.versions_to_tags[version]
-      raise "Unknown version #{version}" unless tag
       @version = version
+      scenarios_present = find_scenarios(:tags => :upgrade_scenario).any?(&matching_version_test)
```

Comment:
As for all upgrade scenarios now we have a separate tag i.e `:upgrade_scenario` so can't we load all scenarios with tag ':upgrade_scenario' in the beginning only?
i.e `find_scenarios(:tags => :upgrade_scenario)`

If we will need to apply any additional conditions like at [line 50](https://github.com/theforeman/foreman_maintain/pull/287/commits/d227bc995cbda0630a6dcf9894412d8e5e668b23?file-filters%5B%5D=.rb&hide-deleted-files=true#diff-3a0557c256d67261f49578f70854cb9fR50) then we already have `filter` method which takes scenarios & conditions as arguments. 
Example - detector.filter(upgrade_scenarios, { :tags => [phase] })

Reply:
@kgaikwad I'm not sure but it seems we are doing that already. Correct  me if I'm wrong:
`find_scenarios` calls `detector.available scenarios` where detector is singleton and the method is defined as singleton too  https://github.com/theforeman/foreman_maintain/blob/master/lib/foreman_maintain/detector.rb#L58-L66. In the end it seems it only calls `detector.filter` as you suggested.

Despite all that I noticed we call the expensive confine block in each upgrade scenario multiple times and that is because we are creating new scenario for each upgrade phase and those share the same confine block.
Here we can spare some cycles by either caching the rpm query calls or making the shared confine a singleton.
I can see much more potential in caching of the rpm queries but it is question if it can't kick us back. Maybe cleaning the cache on calling yum update/install/erase would make it safe. But I'd suggest to address this in a separate PR.

Reply:
@kgaikwad I have checked the code flow and can see available_scenarios is singleton. @mbacovsky I agree to confine block suggestion !

Reply:
> @kgaikwad I'm not sure but it seems we are doing that already. Correct me if I'm wrong:
> `find_scenarios` calls `detector.available scenarios` where detector is singleton and the method is defined as singleton too https://github.com/theforeman/foreman_maintain/blob/master/lib/foreman_maintain/detector.rb#L58-L66. In the end it seems it only calls `detector.filter` as you suggested.


@mbacovsky, you are right about singleton. My point was that we have upgrade scenarios with tag ':upgrade_scenario' and at three places given call to the method `find_scenarios(:tags => :upgrade_scenario)` so would it make sense to extract out this method calling in `upgrade_runner`?


> I can see much more potential in caching of the rpm queries but it is question if it can't kick us back. Maybe cleaning the cache on calling yum update/install/erase would make it safe. But I'd suggest to address this in a separate PR.

+1 for this suggestion.